app/core/auth-options.ts

```diff
@@ -12,7 +14,69 @@ import prisma from '@/core/prisma';
 import { createEvent } from '@/services/db';
 import { upsertUser } from '@/services/db/user';
 
+const USER_TOKEN_REFRESH_MIN = 3; // 3 minutes
+
+interface Token {
+  email: string;
+  accessToken: string;
+  refreshToken: string;
+  idToken: string;
+  roles: string[];
+  sub: string;
+  teams: TeamAccess[];
+  isRefreshTokenExpired: boolean;
+}
+
+interface TeamAccess {
+  clientId: string;
+  roles: string[];
+}
+
+interface DecodedToken {
+  resource_access: Record<string, { roles: string[] }>;
+  sub: string;
+  email: string;
+}
+
+async function processToken(token: Token) {
+  const decodedToken: DecodedToken = jwt.decode(token.accessToken || '') as DecodedToken;
+
+  const newRoles = _get(decodedToken, `resource_access.${AUTH_RESOURCE}.roles`, []);
+  const newSub = decodedToken?.sub ?? '';
+  const newTeams: TeamAccess[] = [];
+
+  _forEach(decodedToken.resource_access ?? {}, (val, key) => {
+    if (key.startsWith(TEAM_SA_PREFIX)) {
+      token.teams.push({ clientId: key, roles: val.roles });
+    }
+  });
+
+  return {
+    isRefreshTokenExpired: false,
+    roles: newRoles,
+    sub: newSub,
+    teams: newTeams,
+  };
+}
+
+async function getNewTokens(refreshToken: string) {
```

Comment:
use `try catch` here to check if the call fails; which means the refresh token is expired.